x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/support/StringMatcher.java

```diff
@@ -0,0 +1,160 @@
+import org.elasticsearch.ElasticsearchSecurityException;
+import org.elasticsearch.common.Strings;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+/**
+ * This class acts as a facade / encapsulation around the expression and testing of string-based patterns within Elasticsearch security.
+ * Security supports "wildcards" in a number of places (e.g. index names within roles). These cases also support
+ * {@link org.apache.lucene.util.automaton.RegExp Lucene-syntax regular expressions} and are implemented via Lucene
+ * {@link org.apache.lucene.util.automaton.Automaton} objects.
+ * However, it can be more efficient to have special handling and avoid {@code Automata} for particular cases such as exact string matches.
+ * This class handles that logic, an provides a clean interface for
+ * <em>test whether a provided string matches one of an existing set of patterns</em> that hides the possible implementation options.
+ */
+public class StringMatcher implements Predicate<String> {
+
+    private static final StringMatcher MATCH_NOTHING = new StringMatcher("(empty))", s -> false);
+
+    private final String description;
+    private final Predicate<String> predicate;
+    private static final Logger LOGGER = LogManager.getLogger(StringMatcher.class);
+
+    private StringMatcher(String description, Predicate<String> predicate) {
+        this.description = description;
+        this.predicate = predicate;
+    }
+
+    public static StringMatcher of(Iterable<String> patterns) {
+        return StringMatcher.builder().includeAll(patterns).build();
+    }
+
+    public static StringMatcher of(String... patterns) {
+        return StringMatcher.builder().includeAll(patterns).build();
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    @Override
+    public String toString() {
+        return description;
+    }
+
+    @Override
+    public boolean test(String s) {
+        return predicate.test(s);
+    }
+
+    public StringMatcher or(Predicate<? super String> other) {
+        Objects.requireNonNull(other);
+        return new StringMatcher(description + "|" + other, this.predicate.or(other));
+    }
+
+    public StringMatcher and(Predicate<? super String> other) {
+        return this.and(String.valueOf(other), other);
+    }
+
+    public StringMatcher and(String otherDescription, Predicate<? super String> otherPredicate) {
+        Objects.requireNonNull(otherPredicate);
+        return new StringMatcher(this.description + "&" + otherDescription, this.predicate.and(otherPredicate));
+    }
+
+    public static class Builder {
+        private final List<String> allText = new ArrayList<>();
+        private final Set<String> exactMatch = new HashSet<>();
+        private final List<String> nonExactMatch = new ArrayList<>();
```

Comment:
Any reason this is not a `Set`?

Reply:
I've made it a `LinkedHashSet`. Order has a small potential performance impact, and I don't want to break something, but there's definitely no value in duplicates.